[SPARK-27388][SQL] encoder for objects defined by properties (ie. Avro)

mazeboard opened the pull request:
## What changes were proposed in this pull request?

This PR adds expression encoders for beans, java.util.List, java.util.Map and java enum. This addition make it possible to encode Avro objects.

The Beans are objects defined by properties; A property is defined by a setter and a getter functions where the getter return type is equal to the setter unique parameter type and the getter and setter functions have the same name; if the getter name is prefixed by "get" then the setter name must be prefixed by "set"; see tests for bean examples.

Avro objects are beans and thus we can create an expression encoder for avro objects as follows:
```
implicit val exprEncoder = ExpressionEncoder[Foo]()
```
All avro types, including fixed types, and excluding complex union types, are suppported by this addition.

The avro fixed types are beans  with exactly one property: bytes.

Currently complex avro unions are not supported because a complex union is declared as Object and there is no expression encoder for Object type (need to use a custom serializer like kryo for example)

## How was this patch tested?

currently only 1 encodeDecodeTest was added to ExpressionEncoderSuite; the test uses an avro object with all accepted types (ie. array, map, fixed, bytes, enum, ...). 

```
    AvroExample1.newBuilder()
      .setMyarray(List("Foo", "Bar").asJava)
      .setMyboolean(true)
      .setMybytes(java.nio.ByteBuffer.wrap("MyBytes".getBytes()))
      .setMydouble(2.5)
      .setMyfixed(new Magic("magic".getBytes))
      .setMyfloat(25.0F)
      .setMyint(100)
      .setMylong(10L)
      .setMystring("hello")
      .setMymap(Map(
        "foo" -> new java.lang.Integer(1),
        "bar" -> new java.lang.Integer(2)).asJava)
      .setMymoney(Money.newBuilder().setAmount(100.0F).setCurrency(Currency.EUR).build())
      .build(),
    "Avro encoder with map, array and fixed types")(
    ExpressionEncoder[AvroExample1])
```

mazeboard committed:
init avro encoder

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

mazeboard committed:
Add support to inherited properties

with this addition avro fixed types are encoded correctly

Currently the avro complex unions are not supported

mazeboard committed:
add encodeDecodeTest for Avro example object

srowen commented:
Overall: isn't this what the Encoder for Java Beans is for and already supports? I am not sure it obvious to me this should go in ScalaReflection.
Separately I don't think it's worth further supporting JavaBean properties. That's pretty old and rare, I think.

mazeboard commented:
1. JavaBean do not support Avro fixed types, because the fixed type has one
property and is named `bytes`; javaBean only accepts properties fixed with
set/get

2. I believe that the current implementation has a bug, indeed

Line 136
in sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala
should be:
val properties = getJavaBeanReadableAndWritableProperties(other) and not
val properties = getJavaBeanReadableProperties(other) 3. toDS, toDF, always
uses the expression encoder I corrected the bug mentioned above and tested
it, by using Encoders.bean, and I had three issues: 1. the ds.map fails
with no encoder found because it is using the ScalaReflection to match for
an encoder 2. the Encoders.bean fails for java enums (assertion fails, not
a StructType since an enum is saved as String) 3. and for Avro fixed types
are not supported because the property in fixed types is not prefixed by
set/get regards Taoufik

On Mon, Apr 8, 2019 at 3:47 PM Sean Owen <notifications@github.com> wrote:

> Overall: isn't this what the Encoder for Java Beans is for and already
> supports? I am not sure it obvious to me this should go in ScalaReflection.
> Separately I don't think it's worth further supporting JavaBean
> properties. That's pretty old and rare, I think.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_apache_spark_pull_24299-23issuecomment-2D480838251&d=DwMCaQ&c=oYSInPoxeTs4m10ilnGu9Q&r=tJmPZfODPMHyXvtcDPDLh8Y6NivZyWO2Jcs2xoRat90&m=1XuL-1AY_mS1qCc4oZ3D2tNdNu8ktGbSIZBkjEfoOFU&s=03jB8i0A-2N65hOzloj_HlIMEx8b4_TWOuCjXoFMwbg&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AG82a0A2aSqET6khk5U0VMPd4vP6-2D5EJks5ve0hjgaJpZM4cdqpS&d=DwMCaQ&c=oYSInPoxeTs4m10ilnGu9Q&r=tJmPZfODPMHyXvtcDPDLh8Y6NivZyWO2Jcs2xoRat90&m=1XuL-1AY_mS1qCc4oZ3D2tNdNu8ktGbSIZBkjEfoOFU&s=UeN37Pqqn63me2GLDt3OxMo7wZ-0G5_MOFDttZIyKQg&e=>
> .
>

-- 
This e-mail and any attachment are confidential and intended solely for the 
use of the individual to whom it is addressed. If you are not the intended 
recipient, please telephone or email the sender and delete this message and 
any attachment from your system. Unauthorized publication, use, 
dissemination, forwarding, printing or copying of this e-mail and its 
associated attachments is strictly prohibited.



http://disclaimer.carrefour.com/ <http://disclaimer.carrefour.com/>


Let's 
respect the environment together. Only print this message if necessary.

mazeboard commented:
@srowen This PR is mainly about adding support to Avro objects and make our code easier

gatorsmile commented:
cc @gengliangwang @cloud-fan

bdrillard commented:
Many thanks to @mazeboard for bringing this PR to my attention and for taking a crack at the problem of typed Avro Datasets in Spark!

I feel it's a matter of due diligence for me to point to another PR supporting Avro-typed Datasets in Spark, namely #22878, which (full transparency) is the work of @xuanyuanking and myself. The approaches taken here and there are different, and it would seem so are the coverage of the Avro spec. I'd like to take the time to compare/contrast.

I am more qualified to speak to the approach and capabilities introduced #22878 (which has a history going back to [Spark-Avro](https://github.com/databricks/spark-avro/pull/217)), and so if I misread this PR in that process, @mazeboard, please do correct my understanding.

I apologize for the length of this reply ahead of time!

## #24299

I'll summarize my reading of this PR's approach: to extend the existing Java `Bean` Encoder functionality to more broadly support tricky types generated by Avro `SpecificRecord` classes, especially Bytes, which Avro doesn't allow access to via typical getters/setters, and Enums, which have to be encoded as Strings and therefore have to be “discovered” via their class.

One stated limitation is complex Union types, which Avro will represent as nested Object. It’s stated that there isn’t an Encoder for Object type (I think a serializer/deserializer could _perhaps_ be made using the [Object](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/ObjectType.scala) and [NewInstance](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala#L431 , but I can’t say how it would work in this Reflection-driven approach). It’s said things could get tough with how Avro serializes those objects when things like Kyro are used. I can see that as a limitation to this Bean-based approach even if the DataTypes and Expressions were sorted out.

Correct my understanding if this is wrong, but because this approach is based on Reflection over the types of the generated SpecificRecord class (viewing Avro as a Bean), it would not be (in a first-class sense) “aware” of the AvroSchema which generates the Class. I think this distinction may matter, and I’ll discuss it more below.

## #22878

To summarize the approach of #22878: create an AvroEncoder which generates a pair of serializer/deserializer Expressions based on the AvroSchema.

This work stands on the Avro efforts that were recently folded into Spark-proper from the (now deprecated) Databricks/Spark-Avro project, but which still _does not_ provide support for a first-class Encoder, and the efficiency and Strong/Static typing that entails. Being based on Spark-Avro however, #22878 _does_ gain benefits of an AvroSchema driven approach. To avoid confusion, I’m going to refer to this folded-in functionality as “Spark-Avro”, relating to [this](https://github.com/apache/spark/tree/master/external/avro) portion of the current Spark project, rather than to the former deprecated project.

### AvroSchema

Because #22878 generates its Encoder through the AvroSchema, we gain a couple things:

1. An isomorphism between the Dataset Schema and the source of truth for the structure of any Avro, namely, its AvroSchema. The important thing here is we can generate an Encoder both from the Class and the AvroSchema in its different representations, like JSON, which opens support for Schema evolution, and use of Schema stores. 
2. Direct support Encoders of any `GenericRecord`, where an AvroSchema is known, but no `SpecificRecord` class can be generated.

### Coverage

Spark-Avro’s ability to move between AvroSchema and Spark Dataset Schema also gives us the ability to traverse the DataType to create our Encoder’s Ser/De Expressions rather than using Reflection. This gives us two immediate benefits

1. The qualities of Bytes and Enums are more transparently represented in the AvroSchema, and so the SerDe expressions can be more directly generated.
2. Nested Records and Complex Unions (as well as Avro Lists and Maps) are a solved-problem in Spark-Avro, so we can generate a Dataset Schema of arbitrary complexity and nesting.
3. Naturally, we don’t have to extend the Bean Encoder for properties.

These two items mean the AvroEncoder in #22878 can generate an `Encoder` having full coverage of Avro types, and this coverage of the various combinations of types that can appear when adding nested Records and Unions is well tested in the PR.

## Last thoughts

The PR goes a long way in support of Avro while still being very concise, which is definitely advantageous from a maintainability perspective. My concerns with a Reflection-based approach are:

1. It’s unknown or perhaps would prove difficult to extend support for (especially) Union types, which are _very_ common in Avro and the _only_ way to express Nullability in Avro.
2. A Reflection-based approach also foregoes Datasets of `GenericRecord` objects.

Parting words for #22878: 

1. While it’s length (in terms of lines of code) as a feature has been discussed, however, I’d say ultimately it’s very well tested, considering we’re ultimately testing a closed set of type combinations described by AvroSchema, and nested complex types are tested via inductive unit tests. Anecdotally, I’ll say we’ve been using it in a fork of Spark-Avro with great success over exceptionally complicated Schemas.
2. It fits quite natively in the new Spark-Avro “external” sub-project, and accomplishes the goal: providing an Encoder for Datasets typed by arbitrary Avro (Generic and Specific).

Again, I'm very happy to see where this discussion goes as it evolves (:

mazeboard committed:
review comment and rename fieldName to propertyName

mazeboard commented:
This PR is not an extension of the existing Java Bean Encoder: The PR adds support for bean objects, java.util.List, java.util.Map, and java enums to ScalaReflection;  unlike the existing javaBean Encoder, properties can be named without the set/get prefix.

Reminder of Avro types:
   primitive types: null, boolean, int, long, float, double, bytes, string
   complex types: Records, Enums, Arrays, Maps, Unions, Fixed

All Avro types are supported by this PR (may be a better test including all Avro types is required), including simple union types and excluding complex union types: Simple  Avro unions are [null, type1], all other unions are complex.

this PR supports simple unions but not complex unions for the simple reason that the Avro compiler will generate java code with type Object for all complex unions, and fields with simple unions will be typed as the non-null type of the union. 

Currently, the ScalaReflection does not have an encoder for Object type, but we can modify the ScalaReflection to use a default encoder (ie. kryo) for Object type (currently I do not know if this advisable, but why not?).

I do not understand the issue related to being Reflection-driven approach: all common scala objects are encoded using reflection.

I may be wrong, but as I tried to explain in this PR, that we need to add types in ScalaReflection to be able to transform Datasets to other Datasets of embedded Avro types. 

As an example the map function transforms the Dataset[A] to a Dataset[(B, C)]

{code} 
val r: Dataset[A] = List(makeA).toDS() 
val ds: Dataset[(B, C)] = r.map(e => (e.getB, e.getC))
{/code}

The map function will recursively use ScalaReflection to find encoders for B, C types (Do you know if this compiles with #22878 solution? Or does it complain with no encoder found?)

Finally, I did not understand the benefits of AvroSchema driven approach, for me an Avro object is completely defined by its properties (that are derived from the Avro schema); the Avro compiler generates java code with all the properties in the Avro schema.

bdrillard commented:
Thanks @mazeboard for the reply and clarifying the support here for Unions. I've edited my post to just a discussion of Complex Union types for clarity:

> unions of two types where `Null` is not one of the types, or [Unions] having more than two types

I'm inclined to wait a bit for some more direction or questions from the Spark committers.

xuanyuanking commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala:
```diff
@@ -308,6 +309,19 @@ class ExpressionEncoderSuite extends CodegenInterpretedPlanTest with AnalysisTes
   encodeDecodeTest(Option("abc"), "option of string")
   encodeDecodeTest(Option.empty[String], "empty option of string")
 
+  encodeDecodeTest(
```
Thanks @bdrillard for the summary of previous work.

Seems #22878 covered this scenario? Please correct me if I missing something. https://github.com/apache/spark/pull/22878/files#diff-24ca7610c9c163779104e6c797713431R327

mazeboard committed:
1. Add support for Avro bytes type

Currently all Avro types are supported, except complex unions

2. for testing, add an Avro having all possible types:
  primitive types: null, boolean, int, long, float, double, bytes, string
  complex types: Records, Enums, Arrays, Maps, Unions, Fixed

mazeboard commented:
just a note about the number of lines changes: the java code for the Avro objects used in the test have 1252 lines of code; so if we ignore those lines of code the overall changed lines of code is +288 -17

mazeboard commented:
I would like to give an example to explain why the PR addition 
must be in ScalaReflection

The test is done with spark-sql 2.4.0

The first transformation (map), from Dataset[Foo) to Dataset[Bar],
pass without error, but the second transformation, from Dataset[Foo] 
to Dataset[(Foo, Bar)] raises an exception, No Encoder found; this 
is explained by the fact that the tuple encoder will search recursively
in ScalaReflection and fails to find an encoder for Foo and Bar


    import spark.implicits._

    implicit val encoderFoo = Encoders.bean[Foo](classOf[Foo])
    implicit val encoderBar = Encoders.bean[Bar](classOf[Bar])

    val bar = new Bar
    bar.setBar("ok")
    val a = new Foo
    a.setA(55)
    a.setB(bar)
    val ds = List(a).toDS()
    val x = ds.map(_.b)               // Ok
    val y = ds.map(x => (x, x.getB))   // Ko - No Encoder found for Bar and Foo

    class Bar {
        private var bar$: java.lang.String = _
        def setBar(value: java.lang.String): Unit = {
           bar$ = value
        }
        def getBar(): java.lang.String = bar$
     }

     class Foo extends Bar {
          var a: java.lang.Integer = _
          var b: Bar = _
          def getA() = a
          def setA(x: java.lang.Integer) {a = x}
          def getB(): Bar = b
          def setB(x: Bar) { b = x}
     }


With the PR additon the following code runs without errors

    import spark.implicits._

    implicit val encoderFoo = ExpressionEncoder[Foo]

    val bar = new Bar
    bar.setBar("ok")
    val a = new Foo
    a.setA(55)
    a.setB(bar)
    val ds = List(a).toDS()
    val x = ds.map(_.b)               // Ok
    val y = ds.map(x => (x, x.getB))   // Ok

mazeboard commented:
Currently we modified JavaTypeInference to be able to create encoders for Avro objects; we have now two solutions, the one in this PR and another one using Encoders.bean (fewer code changes); which one is better? how to proceed? we should create another PR for the Encoders.bean solution?

mazeboard commented:
we created a new PR (https://github.com/apache/spark/pull/24367) for the modification of the bean encoder (JavaTypeInference)

mazeboard committed:
match getters and setters if both are prefixed by get and set respectively
or both are not prefixed

mazeboard committed:
modify check of correctness in encodeDecodeTest

The issue here is how the input and the convertedBack objects are compared

if we replace the check, in line ExpressionEncoderSuite.scala:442,

left.asInstanceOf[Comparable[Any]].compareTo(right) == 0
By
left.asInstanceOf[Comparable[Any]].equals(right) == 0

the test for Avro encoder passes, but unfortunately other tests fail

Equality of objects is tricky.

The GenericData compare function

  /** Comparison implementation.  When equals is true, only checks for equality,
   * not for order. */
  @SuppressWarnings(value="unchecked")
  protected int compare(Object o1, Object o2, Schema s, boolean equals) {
fails to compare Maps when the parameter equals is false

I propose the following

replace ExpressionEncoderSuite:434:444 lines

      val isCorrect = (input, convertedBack) match {
        case (b1: Array[Byte], b2: Array[Byte]) => Arrays.equals(b1, b2)
        case (b1: Array[Int], b2: Array[Int]) => Arrays.equals(b1, b2)
        case (b1: Array[Array[_]], b2: Array[Array[_]]) =>
          Arrays.deepEquals(b1.asInstanceOf[Array[AnyRef]], b2.asInstanceOf[Array[AnyRef]])
        case (b1: Array[_], b2: Array[_]) =>
          Arrays.equals(b1.asInstanceOf[Array[AnyRef]], b2.asInstanceOf[Array[AnyRef]])
        case (left: Comparable[_], right: Comparable[_]) =>
          left.asInstanceOf[Comparable[Any]].compareTo(right) == 0
        case _ => input == convertedBack
      }
by

      val convertedBackRow = encoder.toRow(convertedBack)
      val isCorrect = row == convertedBackRow
With the proposed modification all the tests in ExpressionEncoderSuite passes

mazeboard committed:
refactor getObjectProperties

mazeboard committed:
equality test first

mazeboard committed:
use startsWith

mazeboard commented:
> Overall: isn't this what the Encoder for Java Beans is for and already supports? I am not sure it obvious to me this should go in ScalaReflection.
> Separately I don't think it's worth further supporting JavaBean properties. That's pretty old and rare, I think.

Hi, I would like to hear from about the possibility to merge this PR, many developer teams need to have the possibility to work with Datasets of SpecificRecords (ie. Avro), I implemented a version that modifies the bean encoder but I believe that adding encoders to ScalaReflection is the way to go for reason I explained in this PR, please let me know if this can be considered as a viable solution
